openwrt/openwrt
# yafut: add a kernel update tool for MikroTik NAND

## Background

Commit 9d96b6fb72 ("ath79/mikrotik: disable building NAND images") disabled building images for Mikrotik devices with NAND flash due to a less than satisfactory method used for updating the kernel on those devices back then.  The log message for the above commit contains a neat summary of the problems that this PR attempts to solve:

```
The current support for MikroTik NAND-based devices relies on a
gross hack that packs the kernel into a static YAFFS stub, as the
stock bootloader only supports booting a YAFFS-encapsulated kernel.

The problem with this approach is that since the kernel partition is
blindly overwritten without any kind of wear or badblock management
(due to lack of proper support for YAFFS in OpenWRT), the NAND flash
is not worn uniformly and eventually badblocks appear, leading to
unbootable devices.

This issue has been reported here [1] and discussed in more detail
here [2].

[1] https://forum.openwrt.org/t/rb433-bad-sector-cannot-start-openwrt/71519
[2] https://github.com/openwrt/openwrt/pull/3026#issuecomment-673597461

Until a proper fix is found (or the stock bootloader supports other
filesystems), we disable building these images to prevent unknowing
users from risking their devices.
```

## How this PR proposes to solve the problem

This PR adds support for updating the kernel on Mikrotik devices with NAND flash using a new tool, [Yafut](https://github.com/kempniu/yafut), which enables copying files from/to Yaffs file systems even if the kernel does not have native support for the Yaffs file system compiled in.  Instead of erasing the entire NAND partition holding the kernel during every system upgrade, Yafut preserves the Yaffs filesystem present on that partition and only replaces the kernel executable.  This allows bad block information to be preserved across sysupgrade runs and also enables wear leveling on the NAND partition holding the kernel.

The catch is that the tool relies on the MEMREAD ioctl for MTD character devices that was only first included in Linux 6.1.  That ioctl allows userspace applications to use the Linux kernel's OOB autoplacement mechanism while reading data from NAND devices.  This PR includes a 5.15 backport of the kernel patch series implementing the MEMREAD ioctl.

## Tests carried out so far

In the [tests I ran myself on a Mikrotik RB951Ui-2HnD router](https://github.com/openwrt/openwrt/pull/12226), Yafut was able to successfully replace both the stock RouterOS kernel when run from an OpenWRT initramfs and an OpenWRT kernel when run from an OpenWRT installation booted from NAND.

I do not own a device with bad blocks on the NAND flash, but I carried out some tests using nandsim that seem to show that Yafut is able to handle these just fine (of course as long as the number of available non-bad blocks on the MTD partition holding the kernel is large enough to fit the kernel file).

## How to test

The simplest way to test the tool on a given device is to try reading the kernel executable from the relevant MTD partition and then, if reading works, try to write the kernel executable back to the MTD, e.g.:

    cd /tmp; yafut -d /dev/mtd1 -r -i kernel -o kernel-copy && yafut -d /dev/mtd1 -w -i kernel-copy -o kernel

If the device boots fine after running the above commands and restarting, then updating the kernel via sysupgrade should work just as well.

If the above invocation is not able to read the `kernel` file from flash, try adding the `-T` command-line option to it.  This forces use of inband tags (see kempniu/yafut#10).  If adding this option enables the tool to read the `kernel` file, make sure to also add it to the invocation used for writing the `kernel` file back to the MTD.

## Other notes

  - ~~This PR does **not** switch any existing Mikrotik device to the new kernel update method as it has not been tested widely yet.  See #12226 for an [example](https://github.com/openwrt/openwrt/pull/12226/commits/3ebd5e91b1d22a684fb328b59856e4b02c806412#diff-1412db68ca9f904f7f6124a6786fecdce34b3afa21843b806241971e9198ac3e) of how to switch a specific device to Yafut.  (I do not own any currently-supported device that I could switch over to the new method immediately in this PR.)~~ This PR switches all MikroTik NAND devices to use Yafut because OpenWRT images for these devices are currently not built by default and this PR intends to change that. (See a discussion [below](#pullrequestreview-1359424572) for more information.)

  - NOR flash is currently not supported, but work is underway to change that.
